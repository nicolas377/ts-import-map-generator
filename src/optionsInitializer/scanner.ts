export const enum ScannerNodeKind {
  Whitespace,
  Text,
}

interface ScannerNode {
  kind: ScannerNodeKind;
  text: string;
}

interface TextNode extends ScannerNode {
  kind: ScannerNodeKind.Text;
}

export interface TextNodeWithFlags extends TextNode {
  dashFlag: boolean;
  equalsFlag: boolean;
}

export interface WhitespaceNode extends ScannerNode {
  kind: ScannerNodeKind.Whitespace;
}

function createWhitespaceNode(text: string): WhitespaceNode {
  return {
    kind: ScannerNodeKind.Whitespace,
    text,
  };
}

function createTextNode(text: string): TextNode {
  return {
    kind: ScannerNodeKind.Text,
    text,
  };
}

function regexMatchesFullString(text: string, regex: RegExp): boolean {
  return regex.exec(text)?.[0] === text;
}

const whitespaceRegex = /\s+/;

// Scanner is a generator that yields raw text and whitespace nodes from the input.
export function* scanArgs(
  args: string
): Generator<WhitespaceNode | TextNodeWithFlags> {
  // The scanner is built around matching phrases and returning them as nodes.
  // When the scanner encounters a phrase, it yields a node, clears the current text, and then continues scanning from the end of the phrase.
  // There are two special characters, and text nodes:
  // - The dash character (-) is a flag that indicates the next text node should be treated as a flag.
  //   One dash indicates a short flag, and two dashes indicate a long flag.
  //   Short flags can be combined, e.g. -abc is equivalent to -a -b -c.
  //   Since short flags can be combined, they cannot have values attached to them.
  // - The equals character (=) is a flag that indicates the next text node should be treated as a value.
  //   Multiple equals characters are supported, and mean the same as one equal character.

  // TODO: support quotes around values

  const argsChars: readonly string[] = [...args];
  const flagDefaults = {
    dashFlag: false,
    equalsFlag: false,
  };

  let currentCharIndex = -1;
  let currentText = "";

  let isOutOfChars = false;

  // When this is true, we're out of characters to scan after this one, and should yield a node now and exit.
  let dashFlag: boolean = flagDefaults.dashFlag;
  let equalsFlag: boolean = flagDefaults.equalsFlag;

  // Actual implementation here.
  while (true) {
    const nodeOrFalse = createNodeFromCurrentText();

    if (nodeOrFalse === false) {
      if (isOutOfChars) {
        // We can't do anything else, so we have to just exit.
        break;
      } else {
        // The current text doesn't match any phrases, so we'll keep scanning.
        setupNextChar();
      }
    } else {
      // The current text matches a phrase, so we'll yield a node, cleanup this phrase, and then keep scanning.
      yield nodeOrFalse;

      if (isOutOfChars) {
        // We can't do anything else, so we have to just exit.
        break;
      } else {
        cleanupCurrentPhrase();
        setFlagDefaults();
      }
    }
  }

  // This is the main logic of the scanner.
  // It takes the current text, and tries to create a full phrase.
  // If it can't, it returns false.
  // If it can create a full phrase, it checks if the phrase is continued in the next character.
  // If it is, it returns false.
  // If it isn't, it returns a node for the phrase.
  function createNodeFromCurrentText():
    | WhitespaceNode
    | TextNodeWithFlags
    | false {
    if (currentText.length === 0) {
      // If there's no current text, we can't create a node.
      return false;
    }

    const attemptedNode = createNode();

    if (nodeCanBeExtended(attemptedNode)) return false;
    else return addFlagsToNode(attemptedNode);

    function createNode(): WhitespaceNode | TextNode {
      if (regexMatchesFullString(currentText, whitespaceRegex)) {
        return createWhitespaceNode(currentText);
      } else if (regexMatchesFullString(currentText, /^-+$/)) {
        // This is when we have a bunch of dashes, like "--".
        // We can generate a text node for this, and set the dash flag.
        dashFlag = true;
        return createTextNode(currentText);
      } else if (currentText === "=") {
        // Equal signs are separators between names and values.
        // We can generate a text node for them, and set the equals flag.
        equalsFlag = true;
        return createTextNode(currentText);
      } else {
        // After all this, we can create a text node for the current text.
        return createTextNode(currentText);
      }
    }

    // Looks at the next character, and checks if the current node can be extended to include it.
    function nodeCanBeExtended(node: WhitespaceNode | TextNode): boolean {
      const nextChar = viewNextChar();

      if (nextChar === undefined) {
        // If there's no next character, we can't extend the node.
        return false;
      }
      const nextCharIsWhitespace = regexMatchesFullString(
        nextChar,
        whitespaceRegex
      );

      if (node.kind === ScannerNodeKind.Whitespace) {
        return nextCharIsWhitespace;
      } else if (nextCharIsWhitespace) {
        // We have to be in a text node, and whitespace ends text nodes.
        return false;
      } else if (dashFlag) {
        // The next character can be part of the current node if it's a dash.
        return nextChar === "-";
      } else if (equalsFlag) {
        // There should be only one equals sign in a node.
        // However, we should still support the case where there are multiple equals signs in a row.
        return nextChar === "=";
      } else {
        // This is the case where we're in a non-flagged text node, and the next character isn't whitespace.
        // We can extend the node if the next character is a letter, number, dash, or underscore.
        return regexMatchesFullString(nextChar, /[a-zA-Z0-9_-]/);
      }
    }

    function addFlagsToNode(
      attemptedNode: WhitespaceNode | TextNode
    ): WhitespaceNode | TextNodeWithFlags {
      if (attemptedNode.kind === ScannerNodeKind.Whitespace) {
        // Whitespace nodes don't have flags.
        return attemptedNode;
      } else {
        return {
          ...attemptedNode,
          dashFlag,
          equalsFlag,
        };
      }
    }
  }

  function viewNextChar(): string | undefined {
    return argsChars[currentCharIndex + 1];
  }

  function setupNextChar(): void {
    if (isOutOfChars) return;

    const nextCharIndex = currentCharIndex + 1;

    // If this is the last character, set the isOutOfChars flag.
    if (nextCharIndex >= argsChars.length - 1) {
      isOutOfChars = true;
    }

    currentCharIndex = nextCharIndex;
    currentText += argsChars[currentCharIndex] ?? "";
  }

  // Called after a node is yielded, to reset flag values to their defaults before we start searching for the next phrase.
  function setFlagDefaults(): void {
    dashFlag = flagDefaults.dashFlag;
    equalsFlag = flagDefaults.equalsFlag;
  }

  function cleanupCurrentPhrase(): void {
    currentText = "";
  }
}
