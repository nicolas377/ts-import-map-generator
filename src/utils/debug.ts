type JSONable =
  | string
  | number
  | boolean
  | undefined
  | JSONable[]
  | { [key: string]: JSONable };

interface IDebug {
  logs: Log[];

  error(...message: JSONable[]): void;
  warn(...message: JSONable[]): void;
  info(...message: JSONable[]): void;
  log(...message: JSONable[]): void;
  debug(...message: JSONable[]): void;
  trace(...message: JSONable[]): void;
}

export interface Log {
  dateNow: number;
  level: DebugLevel;
  message: string;
}

const enum DebugLevel {
  Error,
  Warning,
  Info,
  Log,
  Debug,
  Trace,
}

const debugLevelNames: Record<DebugLevel, string> = {
  [DebugLevel.Error]: "ERROR",
  [DebugLevel.Warning]: "WARN",
  [DebugLevel.Info]: "INFO",
  [DebugLevel.Log]: "LOG",
  [DebugLevel.Debug]: "DEBUG",
  [DebugLevel.Trace]: "TRACE",
};

// added by rollup during build
declare const productionBuild: boolean;

export const Debug: IDebug = class Debug {
  public static logs: Log[] = [];

  private static createMessageString(...message: JSONable[]): string {
    const stringifyItem = (item: JSONable): string =>
      item === undefined
        ? "undefined"
        : typeof item === "string"
        ? item
        : typeof item === "number"
        ? item.toString()
        : typeof item === "boolean"
        ? item.toString()
        : Array.isArray(item)
        ? `[${item.map(stringifyItem).join(", ")}]`
        : typeof item === "object"
        ? `{${Object.entries(item)
            .map(([key, value]) => `${key}: ${stringifyItem(value)}`)
            .join(", ")}}`
        : item;

    return message.map(stringifyItem).join(" ");
  }

  private static logMessage(level: DebugLevel, ...message: JSONable[]): void {
    if (productionBuild === false) {
      // We need to stringify the message before we log it, because if we don't, parts of the message could be altered.
      Debug.logs.push({
        dateNow: Date.now(),
        level,
        message: Debug.createMessageString(...message),
      });
    }
  }

  public static error(...message: JSONable[]): void {
    Debug.logMessage(DebugLevel.Error, ...message);
  }

  public static warn(...message: JSONable[]): void {
    Debug.logMessage(DebugLevel.Warning, ...message);
  }

  public static info(...message: JSONable[]): void {
    Debug.logMessage(DebugLevel.Info, ...message);
  }

  public static log(...message: JSONable[]): void {
    Debug.logMessage(DebugLevel.Log, ...message);
  }

  public static debug(...message: JSONable[]): void {
    Debug.logMessage(DebugLevel.Debug, ...message);
  }

  public static trace(...message: JSONable[]): void {
    Debug.logMessage(DebugLevel.Trace, ...message);
  }
};

export function quoteString(string: string): string {
  return `"${string}"`;
}
