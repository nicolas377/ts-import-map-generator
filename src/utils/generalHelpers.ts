export function keysOfObject<T extends PropertyKey>(
  obj: Record<T, unknown>
): T[] {
  return Object.keys(obj) as T[];
}

export function entriesOfObject<T extends PropertyKey, U>(
  obj: Record<T, U>
): [T, U][] {
  return Object.entries(obj) as [T, U][];
}

export function objectHasProperty<T extends PropertyKey>(
  record: Readonly<Record<T, unknown>>,
  key: PropertyKey
): key is T {
  return Object.prototype.hasOwnProperty.call(record, key);
}

export function arrayIncludesValue<T>(
  arr: readonly T[],
  value: unknown
): value is T {
  return arr.includes(value as T);
}

export function getElementInArray<T>(array: T[], index: number): T | undefined {
  if (index < 0 || index >= array.length) {
    return undefined;
  }
  return array[index];
}
