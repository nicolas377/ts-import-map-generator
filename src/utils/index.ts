export * from "utils/debug";
export * from "utils/generalHelpers";
export * from "utils/options";
