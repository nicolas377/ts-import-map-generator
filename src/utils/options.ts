import { OptionType } from "./types";

interface IOptions {
  printHelpAndExit: boolean;
  printVersionAndExit: boolean;
}

class OptionsClass implements IOptions {
  [OptionType.Help] = false;
  [OptionType.Version] = false;

  set printHelpAndExit(value: boolean) {
    this[OptionType.Help] = value;
  }
  get printHelpAndExit() {
    return this[OptionType.Help];
  }

  set printVersionAndExit(value: boolean) {
    this[OptionType.Version] = value;
  }
  get printVersionAndExit(): boolean {
    return this[OptionType.Version];
  }
}

export const programOptions = new OptionsClass();
