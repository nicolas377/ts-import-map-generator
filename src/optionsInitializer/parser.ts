import {
  TextNodeWithFlags as ParserTextNode,
  WhitespaceNode as ParserWhitespaceNode,
  scanArgs,
  ScannerNodeKind,
} from "./scanner";
import { Debug } from "utils";

const enum NodeKind {
  SyntaxTree,
  Whitespace,
  Argument,
  Flag,
  Value,
  EqualsSign,
  Dash,
  UnknownText,
}

const enum NodeFlags {
  None = 0,
  HasMoreThanTwoDashes = 1 << 0,
  HasMoreThanOneEquals = 1 << 1,
}

type NodeId = number;

type ExcludeParentFromNode<T extends Node> = Omit<T, "parent">;

interface ParentLessNode {
  readonly kind: NodeKind;
  readonly id: NodeId;
  flags: NodeFlags;
}

interface Node extends ParentLessNode {
  parent: ParentLessNode;
}

interface UnknownTextNode extends Node {
  readonly kind: NodeKind.UnknownText;
  readonly text: string;
}

interface SeparatorNode extends Node {
  readonly kind: NodeKind.EqualsSign;
}

interface ValueNode extends Node {
  readonly kind: NodeKind.Value;
  readonly text: string;
}

interface FlagNode extends Node {
  readonly kind: NodeKind.Flag;
  readonly text: string;
}

interface DashNode extends Node {
  readonly kind: NodeKind.Dash;

  readonly singleDash: boolean;
  readonly doubleDash: boolean;
}

interface WhitespaceNode extends Node {
  readonly kind: NodeKind.Whitespace;
}

interface ArgumentNode extends Node {
  readonly kind: NodeKind.Argument;

  readonly dash: DashNode;
  readonly flag: FlagNode;
  separator?: SeparatorNode;
  value?: ValueNode;
}

interface SyntaxTree extends ParentLessNode {
  readonly kind: NodeKind.SyntaxTree;
  readonly arguments: readonly ArgumentNode[];
}

interface SyntaxTreeApi {
  readonly syntaxTree: SyntaxTree;
  findNodeById(id: NodeId): Node | undefined;
}

export function parseSyntaxTreeFromArgsString(
  argsString: string
): SyntaxTreeApi {
  // The parser is made to construct a syntax tree.
  // The syntax tree starts with a root node, and then has a series of argument nodes.
  // Each argument node has a dash node, a flag node, an equals sign node, and a value node.
  // The parser is built around parsing those nodes from the phrases yielded by the scanner, and creating the syntax tree from them.
  // There are a few special cases:
  // - If the current dash node has one dash, and the scanner yields an unflagged text node, the the text node should be treated with each individual character as a flag.
  //   This is because short flags can be combined, e.g. -abc is equivalent to -a -b -c.
  // - If the current dash node has one dash, a equals sign node and a value node should be ignored.
  //   This is because short flags cannot have values attached to them.

  const idNodeMap = new Map<NodeId, Node>();
  const syntaxTree: SyntaxTree = {
    ...createParentLessNode(NodeKind.SyntaxTree),
    arguments: [],
  };

  const flagDefaults = {
    singleDash: false,
    doubleDash: false,
    moreThanTwoDashes: false,
  };

  let singleDashFlag: boolean = flagDefaults.singleDash;
  let doubleDashFlag: boolean = flagDefaults.doubleDash;
  let moreThanTwoDashesFlag: boolean = flagDefaults.moreThanTwoDashes;

  let currentDashNode: ExcludeParentFromNode<DashNode> | undefined;
  let currentFlagNode: ExcludeParentFromNode<FlagNode> | undefined;
  let currentSeparatorNode: ExcludeParentFromNode<SeparatorNode> | undefined;
  let currentValueNode: ExcludeParentFromNode<ValueNode> | undefined;

  loopThroughScannedNodes((node, { peekNextNode }) => {
    if (node.kind === ScannerNodeKind.Whitespace) {
      if (doubleDashFlag) {
        // We're in the middle of parsing a long flag argument.
      }
      // We don't care about whitespace when not parsing a long flag argument.
      else return;
    } else {
      if (node.dashFlag) {
        if (node.text.length === 1) {
          singleDashFlag = true;
        } else {
          if (node.text.length >= 2) {
            doubleDashFlag = true;
          }
          if (node.text.length > 2) {
            moreThanTwoDashesFlag = true;
          }
        }

        currentDashNode = createParentLessDashNode(node);
      }
    }
  });

  return createApi(syntaxTree);

  interface ScannedNodeLoopHelpers {
    peekNextNode(): ParserTextNode | ParserWhitespaceNode | undefined;
  }

  function loopThroughScannedNodes(
    cb: (
      node: ParserWhitespaceNode | ParserTextNode,
      helpers: ScannedNodeLoopHelpers
    ) => void
  ): void {
    const scannedArgs = [...scanArgs(argsString)];

    for (const [index, node] of scannedArgs.entries()) {
      cb(node, {
        peekNextNode(): ParserTextNode | ParserWhitespaceNode | undefined {
          return scannedArgs[index + 1];
        },
      });
    }
  }

  function resetForNewArgument(): void {
    currentDashNode = undefined;
    currentFlagNode = undefined;
    currentSeparatorNode = undefined;
    currentValueNode = undefined;

    singleDashFlag = flagDefaults.singleDash;
    doubleDashFlag = flagDefaults.doubleDash;
    moreThanTwoDashesFlag = flagDefaults.moreThanTwoDashes;
  }

  function createParentLessDashNode(
    node: ParserTextNode
  ): ExcludeParentFromNode<DashNode> {
    return {
      ...createParentLessNode(NodeKind.Dash),
      flags: moreThanTwoDashesFlag
        ? NodeFlags.HasMoreThanTwoDashes
        : NodeFlags.None,
      singleDash: singleDashFlag,
      doubleDash: doubleDashFlag,
    };
  }

  function createParentLessArgumentNode(
    dash: ExcludeParentFromNode<DashNode>,
    flag: ExcludeParentFromNode<FlagNode>,
    separator?: ExcludeParentFromNode<SeparatorNode>,
    value?: ExcludeParentFromNode<ValueNode>
  ): ArgumentNode {
    const argumentNode: ArgumentNode = {
      ...createParentLessNode(NodeKind.Argument),
      parent: syntaxTree,
      dash: dash as DashNode,
      flag: flag as FlagNode,
    };

    (dash as DashNode).parent = argumentNode;
    (flag as FlagNode).parent = argumentNode;

    if (value) {
      if (separator) {
        argumentNode.separator = { ...separator, parent: argumentNode };
      }
      argumentNode.value = { ...value, parent: argumentNode };
    }

    return argumentNode;
  }

  function createParentLessNode<T extends NodeKind>(
    kind: T
  ): ParentLessNode & { kind: T } {
    return {
      kind,
      flags: NodeFlags.None,
      id: createNodeId(),
    };
  }

  function createApi(syntaxTree: SyntaxTree): SyntaxTreeApi {
    return {
      syntaxTree,
      findNodeById(id: NodeId): Node | undefined {
        return idNodeMap.get(id);
      },
    };
  }

  function createNodeId(): NodeId {
    let i = 0;

    while (idNodeMap.has(i)) {
      i++;
    }

    return i;
  }
}
