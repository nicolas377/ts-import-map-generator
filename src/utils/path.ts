import {
  readdirSync as readDirectory,
  readFileSync as readFile,
  statSync,
} from "node:fs";
import {
  relative as getRelativePath,
  resolve as getResolvedPath,
  join as joinPaths,
  isAbsolute as pathIsAbsolute,
} from "node:path/posix";
import { cwd as getCurrentWorkingDir } from "node:process";
import { Debug } from "./debug";

export const enum PathType {
  DoesNotExist = 1 << 0,
  File = 1 << 1,
  Directory = 1 << 2,
}

// It doesn't matter what the relativeTo is, just that it's consistent.
const relativeTo = getCurrentWorkingDir();

interface IPathData {
  expandedPath: string;
  relativePath: string;
  type: PathType;
  exists: boolean;
}

interface IPath {
  type: PathType;
  exists: boolean;
  readFile(): string;
  readDir(): Path[];
}

function getPathData(path: string): IPathData {
  const fileStats = statSync(path);

  let isDirectory = false;
  const pathExists =
    fileStats.isFile() || (fileStats.isDirectory() && (isDirectory = true));
  const expandedPath = pathIsAbsolute(path)
    ? getResolvedPath(path)
    : getResolvedPath(relativeTo, path);
  const relativePath = getRelativePath(relativeTo, expandedPath);

  return {
    expandedPath,
    relativePath,
    type: getTypeOfFile(),
    exists: pathExists,
  };

  function getTypeOfFile(): PathType {
    if (!pathExists) {
      return PathType.DoesNotExist;
    }
    if (isDirectory) {
      return PathType.Directory;
    }
    return PathType.File;
  }
}

export class Path implements IPath {
  public readonly type: PathType;
  public readonly exists: boolean;
  public readonly relativePath: string;
  public readonly expandedPath: string;

  constructor(path: string) {
    // Empty path is not allowed
    Debug.assert(path !== "", "Path cannot be empty");

    const pathData = getPathData(path);

    this.expandedPath = pathData.expandedPath;
    this.type = pathData.type;
    this.exists = pathData.exists;
    this.relativePath = pathData.relativePath;
  }

  public readFile(): string {
    // checking bitflags to ensure that the path is a file
    Debug.assert(
      this.type & (PathType.DoesNotExist | PathType.Directory),
      "Path is not a file"
    );

    return readFile(this.expandedPath, "utf8");
  }

  public readDir(): Path[] {
    // checking bitflags to ensure that the path is a directory
    Debug.assert(
      this.type & (PathType.DoesNotExist | PathType.File),
      "Path is not a directory"
    );

    const paths: Path[] = [];
    // these strings are paths in the directory
    const directoryContents = readDirectory(this.expandedPath);

    for (const path of directoryContents) {
      const realPath = joinPaths(this.expandedPath, path);
      paths.push(new Path(realPath));
    }

    return paths;
  }
}
