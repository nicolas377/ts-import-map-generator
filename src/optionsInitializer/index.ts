import { argv } from "node:process";
import { parseSyntaxTreeFromArgsString } from "./parser";

export function initializeOptionsFromCliArgs(): void {
  const argsString: string = argv.slice(2).join(" ");
  const testString =
    "--name=value --name value --bool-name= --bool-name -b --name=";

  const syntaxTree = parseSyntaxTreeFromArgsString(testString);
  console.log(syntaxTree);
}
