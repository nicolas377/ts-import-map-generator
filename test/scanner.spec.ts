import { removeDateFromLog } from "./helpers";
import { scanArgs } from "optionsInitializer/scanner";
import { Debug } from "utils";

describe.each([
  "--name=value --name value --bool-name= --bool-name -b --name=",
  "-a -b -c -d -e -f -g -h -i -j",
  "-abc",
])("scanner %p", (argString) => {
  test("Syntax tree snapshot", () => {
    const syntaxTree = [...scanArgs(argString)];

    expect(syntaxTree).toMatchSnapshot();
  });

  test("Log snapshot", () => {
    // just running the generator all the way through
    [...scanArgs(argString)];

    expect(Debug.logs.map(removeDateFromLog)).toMatchSnapshot();
  });
});
