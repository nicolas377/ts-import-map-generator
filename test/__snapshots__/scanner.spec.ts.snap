// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scanner "--name=value --name value --bool-name= --bool-name -b --name=" Log snapshot 1`] = `[]`;

exports[`scanner "--name=value --name value --bool-name= --bool-name -b --name=" Syntax tree snapshot 1`] = `
[
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "--",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "name",
  },
  {
    "dashFlag": false,
    "equalsFlag": true,
    "kind": 1,
    "text": "=",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "value",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "--",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "name",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "value",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "--",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "bool-name",
  },
  {
    "dashFlag": false,
    "equalsFlag": true,
    "kind": 1,
    "text": "=",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "--",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "bool-name",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "b",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "--",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "name",
  },
  {
    "dashFlag": false,
    "equalsFlag": true,
    "kind": 1,
    "text": "=",
  },
]
`;

exports[`scanner "-a -b -c -d -e -f -g -h -i -j" Log snapshot 1`] = `[]`;

exports[`scanner "-a -b -c -d -e -f -g -h -i -j" Syntax tree snapshot 1`] = `
[
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "a",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "b",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "c",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "d",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "e",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "f",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "g",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "h",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "i",
  },
  {
    "kind": 0,
    "text": " ",
  },
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "j",
  },
]
`;

exports[`scanner "-abc" Log snapshot 1`] = `[]`;

exports[`scanner "-abc" Syntax tree snapshot 1`] = `
[
  {
    "dashFlag": true,
    "equalsFlag": false,
    "kind": 1,
    "text": "-",
  },
  {
    "dashFlag": false,
    "equalsFlag": false,
    "kind": 1,
    "text": "abc",
  },
]
`;
